[14.0][IMP] mis_builder_cash_flow: Improve treatment of forecast and layout (WIP)

JordiBForgeFlow opened the pull request:
The problem with the existing MIS Cash Flow report is that the forecasted amount is always considered. For example, when you forecast to receive 100 and then have an invoice for 100, you don't want to see incoming 200, but 100 instead.

Also, when you have forecasted 100 with date 1/1/2023, and now you are in 1/2/2023, the 100 that you forecasted should be gone, and they are now considered in the "Current" column. By 1/2/2023 you are supposed to have the 100 in your bank.

We avoid counting two times for the same amount. One when the amount is forecasted, and then when you either have a receivable due in this date, or the money has already been received in the bank.

With this PR, the layout of the MIS report now separates IN and OUT regardless if it is receivable and payable.

The forecasted amount presented is the original forecasted amount **substracted** with the actual amount from receivables and payables.

Now the report ignores the forecasts that are prior to today's date, as including them would cause an incorrect initial balance, which needs to be your current financial situation without any forecast.

In the example below we created a forecasted receivable in +1W of 600.000 and a forecasted payable in +1W of -200.000.

![image](https://user-images.githubusercontent.com/7683926/215452741-ec14e307-4ea3-4630-a2ef-bff342bb9421.png)

As you can see the IN TOTAL shows 600.000, even when the "In receivable" is 313,375. This is because we forecasted 600.000 for +1W, and we have received already 313.375 of it. We still estimate that we will end up receiving 600.000. That is why the row "In forecast (remaining)" shows 286.625 (that is, the original forecast of 600.000 minus the amount in receivables).

![image](https://user-images.githubusercontent.com/7683926/215452522-c7b81fcd-596a-4b43-953c-2390cb4298e1.png)

pedrobaeza commented:
Please split the changes in several commits according the fix / imp. Even in different PRs for discussing them separately and faster merge.

JordiBForgeFlow commented:
@pedrobaeza All changes are related with the same concept. They cannot be split, because it is intended to change a behaviour that is currently wrong.

pedrobaeza commented:
To separate the layout in IN and OUT is not a fix and it's not related. In fact, this PR should be [FIX], not [IMP].

JordiBForgeFlow commented:
@pedrobaeza Separating IN and OUT is related in fact. Because it is a necessary prerequisite to computing correctly the "In forecasted (remaining)" amount. (and out)

pedrobaeza commented:
OK, but you can still separate that "pre-requisite" in a `[IMP]` commit, and then put the `[FIX]` commit with each code correctly separated to be analyzed.

JordiBForgeFlow commented:
cc @enricostano you may be interested in this

JordiBForgeFlow committed:
[IMP] mis_builder_cash_flow: Separate IN and OUT for liquidity, receivables/payables and forecasts.
In cash flow you should have incoming and outgoing flows, to begin with, and then provide
detail about what is IN and what is OUT, ordering by the probability that this event occrurs.
That is, liquidity first, then receivables/payables, then forecast.

JordiBForgeFlow committed:
[FIX] mis_builder_cash_flow: Add forecasted remaining quantity

JordiBForgeFlow commented:
I have now split into separate commits, and also added the liquidity part inside the IN TOTAL and OUT TOTAL

![image](https://user-images.githubusercontent.com/7683926/215478473-a81cd189-8173-4042-b443-088d2a80273f.png)

However I can see a fundamental flay with the "BALANCE" row. If you have forecasts, the amount will never be correct as it will double count the amount. So, now the amount from the forecast is not double counted in the "IN TOTAL" AND "OUT TOTAL", but the BALANCE is incorrect.

So, we need a way to make sure that the 'BALANCE' column starts with the total bank amount, and in each period it shows the total of the prior period + the period balance. But this is not supported in MIS Builder, AFAIK

pedrobaeza commented:
I don't get why it counts twice in the balance row. Can you please explain?

JordiBForgeFlow commented:
@pedrobaeza It counts now twice, and it counted twice before this change. The fact is that the forecasted lines, when added, they will always count twice in the existing report. In this PR I managed to remove them in the IN and OUT sections, but not in the BALANCE.

I'm afraid that MIS Builder is not the right tool to build a cash flow report, for this reason.

enricostano commented:
Hi @JordiBForgeFlow !

> I'm afraid that MIS Builder is not the right tool to build a cash flow report, for this reason.

Yeah, we're still struggling with this... how Odoo Enterprise solves this?

JordiBForgeFlow commented:
Closing, as I'm planning to implement a liquidity forecast report.
